Detectors/ITSMFT/ITS/workflow/src/DCSAdaposParserSpec.cxx

```diff
@@ -0,0 +1,219 @@
+{
+  this->mSelfName = o2::utils::Str::concat_string(ChipMappingITS::getName(), "ITSDCSAdaposParser");
+}
+
+//////////////////////////////////////////////////////////////////////////////
+void ITSDCSAdaposParser::init(InitContext& ic)
+{
+  LOGF(info, "ITSDCSAdaposParser init...", mSelfName);
+
+  this->mCcdbUrl = ic.options().get<std::string>("ccdb-url");
+
+  this->mVerboseOutput = ic.options().get<bool>("verbose");
+
+  // Read alpide param object from ccdb: this is the first read, object will be refreshed in run()
+  this->mCcdbFetchUrl = ic.options().get<std::string>("ccdb-fetch-url");
+  getCurrentCcdbAlpideParam();
+
+  return;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Main running function
+// Get Data from ADAPOS and prepare CCDB objects
+void ITSDCSAdaposParser::run(ProcessingContext& pc)
+{
+
+  // Retrieve adapos data and process them
+  auto dps = pc.inputs().get<gsl::span<DPCOM>>("input");
+  process(dps);
+
+  if (doStrobeUpload && dps.size() > 0) {
+    // upload to ccdb
+    pushToCCDB(pc);
+    // refresh the local alpide param object
+    getCurrentCcdbAlpideParam();
+  }
+
+  // clear memory
+  mDPstrobe.clear();
+
+  return;
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Function to retrieve AlpideParam object from CCDB
+void ITSDCSAdaposParser::getCurrentCcdbAlpideParam()
+{
+  long int ts = o2::ccdb::getCurrentTimestamp();
+  LOG(info) << "Getting AlpideParam from CCDB url " << mCcdbFetchUrl << " with timestamp " << ts;
+  auto& mgr = o2::ccdb::BasicCCDBManager::instance();
+  mgr.setURL(mCcdbFetchUrl);
+  mgr.setTimestamp(ts);
+  mCcdbAlpideParam = mgr.get<o2::itsmft::DPLAlpideParam<o2::detectors::DetID::ITS>>("ITS/Config/AlpideParam");
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Function to process DPs
+void ITSDCSAdaposParser::process(const gsl::span<const DPCOM> dps)
+{
+
+  // first we check which DPs are missing - if some are, it means that
+  // the delta map was sent
+  if (mVerboseOutput) {
+    LOG(info) << "\n\n\nProcessing new TF\n-----------------";
+  }
+
+  // Process all DPs, one by one
+  for (const auto& it : dps) {
+    processDP(it);
+  }
+
+  /**************************************
+     decide whether to upload or not the object to ccdb. The logic for strobe length is:
+     - compare the value which arrived from ADAPOS with the one stored in CCDB
+     - if the values are different, store it and create a new object for ccdb
+     - refresh the local ccdb object (int run() method)
+     - the logic continue... memory is cleaned at the end of every run() cycle
+  ***************************************/
+  auto mapel = mDPstrobe.begin();
+  if (!mDPstrobe.size()) {
+    doStrobeUpload = false;
+    return;
+  }
+
+  if (mapel->second.payload_pt1 + 8 != mCcdbAlpideParam->roFrameLengthInBC) {
+    mStrobeToUpload = mapel->second.payload_pt1;
```

Comment:
To be sure: you are picking the 1st entry seen, essentially for a random stave. How it will behave for the ROF ramp up phase, where different staves may have different ROF lengths? I guess it will keep updating CCDB at high rate with the values which are anyway wrong for the whole ITS?

Reply:
that's true! Very good question: I cross-checked with Markus and via ADAPOS we ship only the target strobe length and not the one which is changed during the trigger ramp. They are two separate values in DCS.